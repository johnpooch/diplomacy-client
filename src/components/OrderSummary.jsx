import React from 'react';
import styled from '@emotion/styled';

const StyledP = styled.p`
  .aux,
  .piece,
  .source,
  .target {
    text-transform: capitalize;
    font-weight: bold;
  }

  .action {
    font-style: italic;
  }

  span:not(:last-of-type):after {
    content: ' ';
  }
`;

const OrderSummary = (props) => {
  const { order } = props;
  const { aux, source, target, type, piece_type: pieceType } = order;

  let orderSummary = null;

  if (type === 'build') {
    orderSummary = (
      <StyledP>
        <span className="action">{type}</span>
        <span className="piece">{pieceType}</span>
        in <span className="source">{source.name}</span>
      </StyledP>
    );
  }

  switch (type) {
    case 'hold':
      orderSummary = (
        <StyledP>
          <span className="piece">{source.piece.type}</span>
          in <span className="source">{source.name}</span>
          to <span className="action">{type}</span>
        </StyledP>
      );
      break;

    case 'move':
    case 'retreat':
      orderSummary = (
        <StyledP>
          <span className="piece">{source.piece.type}</span>
          in <span className="source">{source.name}</span>
          to <span className="action">{type}</span>
          to <span className="target">{target.name}</span>
        </StyledP>
      );
      break;

    case 'convoy':
    case 'support':
      orderSummary = (
        <StyledP>
          <span className="piece">{source.piece.type}</span>
          in <span className="source">{source.name}</span>
          to <span className="action">{type}</span>
          <span className="aux">{aux.name}</span>
          to <span className="target">{target.name}</span>
        </StyledP>
      );
      break;
    default:
      break;
  }
  return orderSummary;
};

export default OrderSummary;
