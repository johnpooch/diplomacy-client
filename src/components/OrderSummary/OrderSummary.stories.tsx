/* eslint-disable import/no-extraneous-dependencies */
/* eslint-disable react/jsx-props-no-spreading */
import { Meta, Story } from '@storybook/react';
import React, { ComponentProps } from 'react';

import { OrderType, PieceType } from '../../game/types';

import Component from './OrderSummary';

export default {
  title: 'OrderSummary',
  component: Component,
} as Meta;

const Template: Story<ComponentProps<typeof Component>> = (args) => (
  <Component {...args} />
);

export const ArmyHold: Story = Template.bind({});
ArmyHold.args = {
  order: {
    source: 'Venice',
    pieceType: PieceType.ARMY,
    orderType: OrderType.HOLD,
  },
};

export const FleetHold: Story = Template.bind({});
FleetHold.args = {
  order: {
    source: 'Venice',
    pieceType: PieceType.FLEET,
    orderType: OrderType.HOLD,
  },
};

export const ArmyMove: Story = Template.bind({});
ArmyMove.args = {
  order: {
    source: 'Venice',
    target: 'Trieste',
    pieceType: PieceType.ARMY,
    orderType: OrderType.MOVE,
  },
};

export const FleetMove: Story = Template.bind({});
FleetMove.args = {
  order: {
    source: 'Venice',
    target: 'Trieste',
    pieceType: PieceType.FLEET,
    orderType: OrderType.MOVE,
  },
};

export const FleetMoveTargetCoast: Story = Template.bind({});
FleetMoveTargetCoast.args = {
  order: {
    source: 'Venice',
    target: 'Trieste',
    targetCoast: 'NC',
    pieceType: PieceType.FLEET,
    orderType: OrderType.MOVE,
  },
};

export const ArmyRetreat: Story = Template.bind({});
ArmyRetreat.args = {
  order: {
    source: 'Venice',
    target: 'Trieste',
    pieceType: PieceType.ARMY,
    orderType: OrderType.RETREAT,
  },
};

export const FleetRetreat: Story = Template.bind({});
FleetRetreat.args = {
  order: {
    source: 'Venice',
    target: 'Trieste',
    pieceType: PieceType.FLEET,
    orderType: OrderType.RETREAT,
  },
};

export const FleetRetreatTargetCoast: Story = Template.bind({});
FleetRetreatTargetCoast.args = {
  order: {
    source: 'Venice',
    target: 'Trieste',
    targetCoast: 'NC',
    pieceType: PieceType.FLEET,
    orderType: OrderType.RETREAT,
  },
};

export const ArmySupport: Story = Template.bind({});
ArmySupport.args = {
  order: {
    source: 'Venice',
    aux: 'Adriatic Sea',
    target: 'Trieste',
    pieceType: PieceType.ARMY,
    orderType: OrderType.SUPPORT,
  },
};

export const FleetSupport: Story = Template.bind({});
FleetSupport.args = {
  order: {
    source: 'Venice',
    aux: 'Adriatic Sea',
    target: 'Trieste',
    pieceType: PieceType.FLEET,
    orderType: OrderType.SUPPORT,
  },
};

export const FleetConvoy: Story = Template.bind({});
FleetConvoy.args = {
  order: {
    source: 'Venice',
    aux: 'Adriatic Sea',
    target: 'Trieste',
    pieceType: PieceType.FLEET,
    orderType: OrderType.CONVOY,
  },
};

export const BuildArmy: Story = Template.bind({});
BuildArmy.args = {
  order: {
    source: 'Venice',
    pieceType: PieceType.ARMY,
    orderType: OrderType.BUILD,
  },
};

export const BuildFleet: Story = Template.bind({});
BuildFleet.args = {
  order: {
    source: 'Venice',
    pieceType: PieceType.FLEET,
    orderType: OrderType.BUILD,
  },
};

export const DisbandArmy: Story = Template.bind({});
DisbandArmy.args = {
  order: {
    source: 'Venice',
    pieceType: PieceType.ARMY,
    orderType: OrderType.DISBAND,
  },
};

export const DisbandFleet: Story = Template.bind({});
DisbandFleet.args = {
  order: {
    source: 'Venice',
    pieceType: PieceType.FLEET,
    orderType: OrderType.DISBAND,
  },
};
