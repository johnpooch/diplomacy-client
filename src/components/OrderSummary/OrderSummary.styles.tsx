import { makeStyles } from '@material-ui/core/styles';

export default makeStyles(() => {
  return {
    root: {
      '& span:first-of-type': {
        textTransform: 'capitalize',
      },
      '& .text': {
        display: 'inline',
      },
    },
    aux: {},
    orderType: {},
    prep: {},
    pieceType: {},
    source: {},
    target: {},
    targetCoast: {},
  };
});
