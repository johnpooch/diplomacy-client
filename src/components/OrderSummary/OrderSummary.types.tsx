import { OrderDisplay } from '../../types';

export interface OrderSummaryComponentProps {
  order: OrderDisplay;
}
