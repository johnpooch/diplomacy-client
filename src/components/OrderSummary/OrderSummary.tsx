/* eslint-disable react/jsx-props-no-spreading */
import { Typography } from '@material-ui/core';
import React from 'react';

import { OrderType, PieceType } from '../../game/types';
import { Army, Fleet } from '../Icon';

import useStyles from './OrderSummary.styles';
import { OrderSummaryComponentProps } from './OrderSummary.types';

const Aux = ({ territory }) => {
  const classes = useStyles();
  return (
    <Typography variant="body2" component="span" className={classes.aux}>
      {territory}
    </Typography>
  );
};

const Source = ({ territory }) => {
  const classes = useStyles();
  return (
    <Typography variant="body2" component="span" className={classes.source}>
      {territory}
    </Typography>
  );
};

const Prep = ({ term }) => {
  const classes = useStyles();
  return (
    <Typography variant="body2" component="span" className={classes.prep}>
      {term}
    </Typography>
  );
};

const Target = ({ territory }) => {
  const classes = useStyles();
  return (
    <Typography variant="body2" component="span" className={classes.target}>
      {territory}
    </Typography>
  );
};

const TargetCoast = ({ targetCoast }) => {
  const classes = useStyles();
  return (
    <Typography
      variant="body2"
      component="span"
      className={classes.targetCoast}
    >
      ({targetCoast})
    </Typography>
  );
};

const OrderTypeElement = ({ orderType }) => {
  const classes = useStyles();
  return (
    <Typography variant="body2" component="span" className={classes.orderType}>
      {orderType}
    </Typography>
  );
};

const PieceTypeElement = ({ pieceType }) => {
  const classes = useStyles();
  return (
    <Typography variant="body2" component="span" className={classes.pieceType}>
      {pieceType}
    </Typography>
  );
};

const BuildOrderText = ({ pieceType, source, targetCoast, orderType }) => {
  return (
    <div className="text">
      <OrderTypeElement orderType={orderType} />{' '}
      <PieceTypeElement pieceType={pieceType} /> <Prep term="in" />{' '}
      <Source territory={source} />{' '}
      {targetCoast && <TargetCoast targetCoast={targetCoast} />}{' '}
    </div>
  );
};

const DisbandOrderText = ({ source, orderType }) => {
  return (
    <div className="text">
      <Source territory={source} /> <OrderTypeElement orderType={orderType} />
    </div>
  );
};

const HoldOrderText = ({ source, orderType }) => {
  return (
    <div className="text">
      <Source territory={source} /> <OrderTypeElement orderType={orderType} />
    </div>
  );
};

const MoveOrderText = ({ source, target, targetCoast, orderType }) => {
  return (
    <div className="text">
      <Source territory={source} /> <OrderTypeElement orderType={orderType} />{' '}
      <Prep term="to" /> <Target territory={target} />{' '}
      {targetCoast && <TargetCoast targetCoast={targetCoast} />}{' '}
    </div>
  );
};

const AuxOrderText = ({ aux, source, target, orderType }) => {
  return (
    <div className="text">
      <Source territory={source} /> <OrderTypeElement orderType={orderType} />{' '}
      <Aux territory={aux} /> <Prep term="to" /> <Target territory={target} />
    </div>
  );
};

const OrderTypeMap = {
  [OrderType.MOVE]: MoveOrderText,
  [OrderType.HOLD]: HoldOrderText,
  [OrderType.SUPPORT]: AuxOrderText,
  [OrderType.CONVOY]: AuxOrderText,
  [OrderType.RETREAT]: MoveOrderText,
  [OrderType.BUILD]: BuildOrderText,
  [OrderType.DISBAND]: DisbandOrderText,
};

const OrderSummary: React.FC<OrderSummaryComponentProps> = ({ order }) => {
  const classes = useStyles();
  const { orderType, pieceType } = order;
  const icon = pieceType === PieceType.ARMY ? Army : Fleet;
  const OrderTextElement = OrderTypeMap[orderType];
  return (
    <div className={classes.root}>
      {icon} <OrderTextElement {...order} />
    </div>
  );
};

export default OrderSummary;
