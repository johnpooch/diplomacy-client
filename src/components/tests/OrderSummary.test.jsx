import React from 'react';
import { render } from '@testing-library/react';

import OrderSummary from '../OrderSummary';

const source = {
  piece: {
    type: 'army',
  },
  territory: {
    name: 'london',
  },
};

const target = {
  territory: {
    name: 'wales',
  },
};

test('Build army', () => {
  const order = {
    type: 'build',
    piece_type: 'army',
    source,
  };
  const { asFragment } = render(<OrderSummary order={order} />);
  expect(asFragment()).toMatchSnapshot();
});

test('Build fleet', () => {
  const order = {
    type: 'build',
    piece_type: 'fleet',
    source,
  };
  const { asFragment } = render(<OrderSummary order={order} />);
  expect(asFragment()).toMatchSnapshot();
});

test('Hold army', () => {
  const order = {
    type: 'hold',
    piece_type: null,
    source,
  };
  const { asFragment } = render(<OrderSummary order={order} />);
  expect(asFragment()).toMatchSnapshot();
});

test('Move army', () => {
  const order = {
    type: 'move',
    piece_type: null,
    source,
    target,
  };
  const { asFragment } = render(<OrderSummary order={order} />);
  expect(asFragment()).toMatchSnapshot();
});
