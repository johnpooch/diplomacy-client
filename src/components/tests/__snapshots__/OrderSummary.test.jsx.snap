// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Build army 1`] = `
<DocumentFragment>
  <p
    class="css-8t79ir-StyledP e19maj680"
  >
    <span
      class="action"
    >
      build
    </span>
    <span
      class="piece"
    >
      army
    </span>
    in 
    <span
      class="source"
    >
      london
    </span>
  </p>
</DocumentFragment>
`;

exports[`Build fleet 1`] = `
<DocumentFragment>
  <p
    class="css-8t79ir-StyledP e19maj680"
  >
    <span
      class="action"
    >
      build
    </span>
    <span
      class="piece"
    >
      fleet
    </span>
    in 
    <span
      class="source"
    >
      london
    </span>
  </p>
</DocumentFragment>
`;

exports[`Convoy 1`] = `
<DocumentFragment>
  <p
    class="css-8t79ir-StyledP e19maj680"
  >
    <span
      class="piece"
    >
      army
    </span>
    in 
    <span
      class="source"
    >
      london
    </span>
    to 
    <span
      class="action"
    >
      convoy
    </span>
    <span
      class="aux"
    >
      york
    </span>
    to 
    <span
      class="target"
    >
      wales
    </span>
  </p>
</DocumentFragment>
`;

exports[`Hold 1`] = `
<DocumentFragment>
  <p
    class="css-8t79ir-StyledP e19maj680"
  >
    <span
      class="piece"
    >
      army
    </span>
    in 
    <span
      class="source"
    >
      london
    </span>
    to 
    <span
      class="action"
    >
      hold
    </span>
  </p>
</DocumentFragment>
`;

exports[`Move 1`] = `
<DocumentFragment>
  <p
    class="css-8t79ir-StyledP e19maj680"
  >
    <span
      class="piece"
    >
      army
    </span>
    in 
    <span
      class="source"
    >
      london
    </span>
    to 
    <span
      class="action"
    >
      move
    </span>
    to 
    <span
      class="target"
    >
      wales
    </span>
  </p>
</DocumentFragment>
`;

exports[`Retreat 1`] = `
<DocumentFragment>
  <p
    class="css-8t79ir-StyledP e19maj680"
  >
    <span
      class="piece"
    >
      army
    </span>
    in 
    <span
      class="source"
    >
      london
    </span>
    to 
    <span
      class="action"
    >
      retreat
    </span>
    to 
    <span
      class="target"
    >
      wales
    </span>
  </p>
</DocumentFragment>
`;

exports[`Support 1`] = `
<DocumentFragment>
  <p
    class="css-8t79ir-StyledP e19maj680"
  >
    <span
      class="piece"
    >
      army
    </span>
    in 
    <span
      class="source"
    >
      london
    </span>
    to 
    <span
      class="action"
    >
      support
    </span>
    <span
      class="aux"
    >
      york
    </span>
    to 
    <span
      class="target"
    >
      wales
    </span>
  </p>
</DocumentFragment>
`;